src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/hover/LSPTextHover.java

```diff
@@ -181,29 +185,27 @@ private int getTargetOffset(PsiElement originalElement) {
      * @param offset  the target offset.
      */
     private void initiateHoverRequest(PsiElement element, int offset) {
+        if (this.cancellationSupport != null) {
+            this.cancellationSupport.cancel();
+            this.cancellationSupport = null;
+        }
         PsiDocumentManager manager = PsiDocumentManager.getInstance(element.getProject());
         final Document document = manager.getDocument(element.getContainingFile());
         if (offset != -1 && (this.lspRequest == null || !element.equals(this.lastElement) || offset != this.lastOffset)) {
             this.lastElement = element;
             this.lastOffset = offset;
+            this.cancellationSupport = new CancellationSupport();
             this.lspRequest = LanguageServiceAccessor.getInstance(element.getProject())
                     .getLanguageServers(document, capabilities -> isHoverCapable(capabilities))
                     .thenApplyAsync(languageServers -> // Async is very important here, otherwise the LS Client thread is in
                             // deadlock and doesn't read bytes from LS
                             languageServers.stream()
-                                    .map(languageServer -> {
-                                        try {
-                                            return languageServer.getSecond().getTextDocumentService()
-                                                    .hover(LSPIJUtils.toHoverParams(offset, document)).get();
-                                        } catch (ExecutionException e) {
-                                            LOGGER.warn(e.getLocalizedMessage(), e);
-                                            return null;
-                                        } catch (InterruptedException e) {
-                                            LOGGER.warn(e.getLocalizedMessage(), e);
-                                            Thread.currentThread().interrupt();
-                                            return null;
-                                        }
-                                    }).filter(Objects::nonNull).collect(Collectors.toList()));
+                                    .map(languageServer ->
+                                            cancellationSupport.execute(
```

Comment:
if it runs in async, there's a small risk that at this point cancellationSupport could be null-ed by other methods